### PR TITLE
test: add load test pipeline to nightly job

### PR DESCRIPTION
**Pull request by aramase**
**Reason for Change**:

- Adds load test pipelines to nightly job for `managed` and `standard` mode with AKS clusters

**Requirements**

- [x] squashed commits
- [ ] included documentation
- [ ] added unit tests and e2e tests (if applicable). See [test standard](https://github.com/Azure/aad-pod-identity/blob/master/CONTRIBUTING.md#test-standard) for more details.
- [x] ran `make precommit`

**Issue Fixed**:

fixes https://github.com/Azure/aad-pod-identity/issues/714

**Notes for Reviewers**:

### CODE REVIEW

**Comment by aramase** [comment]
Successful test run for `managed` and `standard` mode - https://dev.azure.com/azure/aad-pod-identity/_build/results?buildId=13257&view=results

Setting the # of pods to 2000 for few runs. We can then make this a larger value after few nightly runs. Current # of agents is 35 with 110 max pods config.

File: .pipelines/load-tests-template.yml [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+        - template: deploy-aks-cluster.yml
+        - script: |
+            # Create assigned identity to be used for test
+            POD_IDENTITY_CLIENT_ID=$(az identity create -n ${POD_IDENTITY_ID_NAME} -g MC_${RESOURCE_GROUP}_${RESOURCE_GROUP}_${LOCATION} --query clientId -otsv)
+            echo "##vso[task.setvariable variable=POD_IDENTITY_CLIENT_ID]${POD_IDENTITY_CLIENT_ID}"
+
+            if [[ $OPERATION_MODE == "managed" ]]; then
+              VMSS_NAME=$(az vmss list -g MC_${RESOURCE_GROUP}_${RESOURCE_GROUP}_$(LOCATION) --query "[].name" -o tsv)
+              echo $VMSS_NAME
```
Comment (chewong) [inline .pipelines/load-tests-template.yml:58]:
```suggestion
              echo "Assigning /subscriptions/$(SUBSCRIPTION_ID)/resourcegroups/MC_${RESOURCE_GROUP}_${RESOURCE_GROUP}_$(LOCATION)/providers/Microsoft.ManagedIdentity/userAssignedIdentities/${POD_IDENTITY_ID_NAME} to ${VMSS_NAME}"
```

File: .pipelines/load-tests-template.yml [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    type: object
+
+jobs:
+  - ${{ each operationMode in parameters.operationModes }}:
+    - job:
+      displayName: ${{ format('load/{0}', operationMode) }}
+      strategy:
+        matrix:
+          test-pod-2000:
+            POD_COUNT: 20
```
Comment (chewong) [inline .pipelines/load-tests-template.yml:12]: nit: `POD_COUNT_PER_DEPLOYMENT` would be a more descriptive name.

File: test/load/deployment-template.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    spec:
+      terminationGracePeriodSeconds: 0
+      initContainers:
+        - name: init
+          image: mcr.microsoft.com/azure-cli
+          imagePullPolicy: IfNotPresent
+          command:
+            - sh
+            - -c
+            - az login --identity --allow-no-subscriptions && az login -i -u $POD_IDENTITY_CLIENT_ID --allow-no-subscriptions
```
[the remark this answers is not in the document]
Comment (aramase) [reply test/load/deployment-template.yml:28]: That's right. It checks the logging and also ensures NMI handles both cases as expected.

**Comment by aramase** [comment]
Latest test run with all the changes and master build - https://dev.azure.com/azure/aad-pod-identity/_build/results?buildId=13307&view=results

File: .pipelines/load-tests-template.yml [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            echo "--- CONFIG ---"
+            cat test/load/config.yml
+            echo "--- DEPLOYMENT ---"
+            cat test/load/deployment.yml
+            echo "--- AZURE IDENTITY ---"
+            cat test/load/azureidentity.yml
+            echo "--- AZURE IDENTITY BINDING ---"
+            cat test/load/azureidentitybinding.yml
+          workingDirectory: "$(system.defaultWorkingDirectory)"
+          displayName: "Get yaml"
```
Comment (chewong) [inline .pipelines/load-tests-template.yml:112]:
```suggestion
          displayName: "Get yaml"

```

File: .pipelines/load-tests-template.yml [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            kubectl -n $NAMESPACE wait --for=condition=Ready --timeout=5m pod -l app.kubernetes.io/component=nmi
+            if [[ "${OPERATION_MODE:-}" == "standard" ]]; then
+              kubectl -n $NAMESPACE wait --for=condition=Ready --timeout=5m pod -l app.kubernetes.io/component=mic
+            fi
+          displayName: "Deploy aad-pod-identity"
+          workingDirectory: "$(system.defaultWorkingDirectory)"
+          env:
+            NAMESPACE: "kube-system"
+            SUBSCRIPTION_ID: $(LOAD_TEST_SUBSCRIPTION_ID)
+            POD_IDENTITY_RESOURCE_GROUP: $(CLUSTER_RESOURCE_GROUP)
```
Comment (chewong) [inline .pipelines/load-tests-template.yml:101]:
```suggestion
            POD_IDENTITY_RESOURCE_GROUP: $(CLUSTER_RESOURCE_GROUP)

```

File: .pipelines/load-tests-template.yml [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+            let NODE_COUNT=${AGENT_COUNT}*${NODE_POOL_COUNT}
+            ./run-e2e.sh cluster-loader2 \
+              --nodes=${NODE_COUNT} \
+              --provider=local \
+              --report-dir="$(perf-tests.repo.path)/_artifacts" \
+              --masterip=${MASTERIP} \
+              --master-internal-ip=${MASTERINTERNALIP} \
+              --testconfig="$(system.defaultWorkingDirectory)/test/load/config.yml"
+          workingDirectory: "$(perf-tests.repo.path)"
+          displayName: "Run clusterloader2 test"
```
Comment (chewong) [inline .pipelines/load-tests-template.yml:123]:
```suggestion
          displayName: "Run clusterloader2 test"

```

File: .pipelines/load-tests-template.yml [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+            kubectl get nodes
+            echo "--- GET NS ---"
+            kubectl get ns
+            echo "--- GET PODS ---"
+            kubectl get pods --all-namespaces
+            echo "--- TOP PODS ---"
+            kubectl top pods -n kube-system
+          workingDirectory: "$(perf-tests.repo.path)"
+          condition: succeededOrFailed()
+          displayName: "Get stats"
```
Comment (chewong) [inline .pipelines/load-tests-template.yml:135]:
```suggestion
          displayName: "Get stats"

```

File: .pipelines/load-tests-template.yml [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        - script: |
+            echo "--- MIC LOGS ---"
+            kubectl logs -n kube-system -l app.kubernetes.io/component=mic --tail 5000
+            echo "--- NMI LOGS ---"
+            kubectl logs -n kube-system -l app.kubernetes.io/component=nmi --tail 5000
+          workingDirectory: "$(perf-tests.repo.path)"
+          condition: succeededOrFailed()
+          displayName: "Get logs"
```
Comment (chewong) [inline .pipelines/load-tests-template.yml:143]:
```suggestion
          displayName: "Get logs"

```

File: .pipelines/load-tests-template.yml [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        - template: build-images.yml
+
+        - script: |
+            az account set -s=$(LOAD_TEST_SUBSCRIPTION_ID)
+          displayName: "Set subscription for load tests"
+
+        - script: |
+            export RESOURCE_GROUP="pi-load-e2e-$(openssl rand -hex 6)"
+            echo "##vso[task.setvariable variable=RESOURCE_GROUP]${RESOURCE_GROUP}"
+          displayName: "Generate resource group name"
```
Comment (chewong) [inline .pipelines/load-tests-template.yml:46]:
```suggestion
          displayName: "Generate resource group name"

```

File: .pipelines/load-tests-template.yml [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+        - template: deploy-aks-cluster.yml
```
Comment (chewong) [inline .pipelines/load-tests-template.yml:47]:
```suggestion
        - template: deploy-aks-cluster.yml

```

**Comment by aramase** [comment]
> Assigned ID addition: 2h12m21.912364199s doesn't look right. It looks like the stat collector just performs additions of all 2000 durations together. Is this behavior intended? If not, we can address it in a separate PR.

No, we should address this in a separate PR. Opened https://github.com/Azure/aad-pod-identity/issues/749 to track it.

File: test/load/config-deployment-template.yml [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+          - basename: deployment
+            objectTemplatePath: deployment.yml
+  - measurements:
+      - Identifier: WaitForRunningDeployments
+        Method: WaitForControlledPodsRunning
+        Params:
+          action: gather
+  - measurements:
+      - Identifier: ResourceUsageSummary
+        Method: ResourceUsageSummary
```
Comment (sozercan) [inline test/load/config-deployment-template.yml:86]: is there a way to surface/visualize resource usage report?

Comment (aramase) [reply test/load/config-deployment-template.yml:86]: Right, now the artifacts are uploaded to drop and we also print out the result for `kubectl top` as part of the pipeline. We could definitely look into a way to auto publish this as a report/visualize the data.

**Pull request merged**